hastexo-guacamole-client/guacamole/client.py

```diff
@@ -0,0 +1,226 @@
+
+        # Client ID
+        self._id = None
+
+        self.logger = guac_logger
+        if logger:
+            self.logger = logger
+
+        if debug:
+            self.logger.setLevel(logging.DEBUG)
+
+    @property
+    def client(self):
+        """
+        Socket connection.
+        """
+        if not self._client:
+            self._client = socket.create_connection(
+                (self.host, self.port), self.timeout)
+            self.logger.debug('Client connected with guacd server (%s, %s, %s)'
+                              % (self.host, self.port, self.timeout))
+
+        return self._client
+
+    @property
+    def id(self):
+        """Return client id"""
+        return self._id
+
+    def close(self):
+        """
+        Terminate connection with Guacamole guacd server.
+        """
+        self.client.close()
+        self._client = None
+        self.connected = False
+        print('Connection closed.')
+
+    def receive(self):
+        """
+        Receive instructions from Guacamole guacd server.
+        """
+        start = 0
+
+        while True:
+            idx = self._buffer.find(INST_TERM.encode(), start)
+            if idx != -1:
+                # instruction was fully received!
+                line = self._buffer[:idx + 1]
+                self._buffer = self._buffer[idx + 1:]
+                self.logger.debug('Received instruction: %s' % line)
+                return line
+            else:
+                start = len(self._buffer)
+                # we are still waiting for instruction termination
+                buf = self.client.recv(BUF_LEN)
+                if not buf:
+                    # No data recieved, connection lost?!
+                    self.close()
+                    self.logger.debug(
+                        'Failed to receive instruction. Closing.')
+                    return None
+                self._buffer.extend(buf)
+
+    def send(self, data):
+        """
+        Send encoded instructions to Guacamole guacd server.
+        """
+        self.logger.debug('Sending data: %s' % data)
+
+        if isinstance(data, str):
+            data = data.encode()
+        self.client.sendall(data)
+
+    def read_instruction(self):
+        """
+        Read and decode instruction.
+        """
+        self.logger.debug('Reading instruction.')
+        return Instruction.load(self.receive())
+
+    def send_instruction(self, instruction):
+        """
+        Send instruction after encoding.
+        """
+        self.logger.debug('Sending instruction: %s' % str(instruction))
+        return self.send(instruction.encode())
+
+    def handshake(self, protocol='ssh', width=1024, height=768, dpi=96,
+                  audio=None, video=None, image=None, **kwargs):
+        """
+        Establish connection with Guacamole guacd server via handshake.
+        """
+        if protocol not in PROTOCOLS and 'connectionid' not in kwargs:
+            self.logger.debug(
```

Comment:
If you're unconditionally raising an error immediately after, this should be `error()`, not `debug()`.